VungleAdapter/src/main/java/com/chartboost/mediation/vungleadapter/VungleAdapter.kt

```diff
@@ -664,153 +524,198 @@ class VungleAdapter : PartnerAdapter {
     ): Result<PartnerAd> {
-        val adm = adms.remove(partnerAd.request.partnerPlacement)
-
-        return suspendCancellableCoroutine { continuation ->
-            fun resumeOnce(result: Result<PartnerAd>) {
-                if (continuation.isActive) {
-                    continuation.resume(result)
-                }
+        fun canPlay(canPlay: () -> Boolean, play: () -> Unit): Result<PartnerAd> {
+            return if (canPlay()) {
+                play()
+                Result.success(partnerAd)
+            } else {
+                PartnerLogController.log(SHOW_FAILED)
+                Result.failure(ChartboostMediationAdException(ChartboostMediationError.CM_SHOW_FAILURE_AD_NOT_READY))
             }
+        }
 
-            if (Vungle.canPlayAd(partnerAd.request.partnerPlacement, adm)) {
-                Vungle.playAd(
-                    partnerAd.request.partnerPlacement,
-                    adm,
-                    null,
-                    object : PlayAdCallback {
-                        override fun creativeId(creativeId: String) {
-                            // Ignored.
-                        }
+        onShowSuccess = {
+            PartnerLogController.log(SHOW_SUCCEEDED)
+            Result.success(partnerAd)
+        }
 
-                        override fun onAdStart(id: String) {
-                            PartnerLogController.log(SHOW_SUCCEEDED)
-                            resumeOnce(Result.success(partnerAd))
-                        }
+        onShowError = { baseAd, _ ->
+            PartnerLogController.log(
+                SHOW_FAILED,
+                "Vungle failed to show the fullscreen ad for placement " +
+                        "${baseAd.placementId}."
+            )
+        }
 
-                        @Deprecated("Deprecated by Vungle. Use onAdEnd(String) instead.")
-                        override fun onAdEnd(
-                            id: String,
-                            completed: Boolean,
-                            isCTAClicked: Boolean
-                        ) {
-                        }
+        return when (val ad = partnerAd.ad) {
+            null -> {
+                PartnerLogController.log(SHOW_FAILED, "Ad is null.")
+                Result.failure(ChartboostMediationAdException(ChartboostMediationError.CM_SHOW_FAILURE_AD_NOT_FOUND))
+            }
 
-                        override fun onAdEnd(id: String) {
-                            PartnerLogController.log(DID_DISMISS)
-                            listener?.onPartnerAdDismissed(partnerAd, null)
-                                ?: PartnerLogController.log(
-                                    CUSTOM,
-                                    "Unable to fire onAdEnd for Vungle adapter. Listener is null."
-                                )
-                        }
+            is InterstitialAd -> canPlay(ad::canPlayAd, ad::play)
+            is RewardedAd -> canPlay(ad::canPlayAd, ad::play)
```

Comment:
do InterstitialAd and RewardedAd extend base? I don't think you need to call the respective inherited methods if you can just call the base methods. then you don't need this `::` either.

Reply:
Ah, true. `InterstitialAd` and RewardedAd extend `BaseFullscreenAd`